V4.y

LAShemilt opened the pull request:
This is a first pass refactor to get pyscicat working with v4 of Scicat. We would like to : 
- Provide an easy install that can be versioned
- mark where functions are deprecated and breaking changes occur
- have a ci workflow for testing against the v4.y backend

dylanmcreynolds commented on setup.cfg:
```diff
@@ -4,3 +4,54 @@ style = pep440-post
+[metadata]
+
+name = pyscicat
+version = 1.0.0
```
In the previous version, setup.py got used versioneer to detect the version from git tags. I'm sure this is a solved problem using setup.cfg...maybe just remove this?

LAShemilt replied:
so I use bump2version for automatic versioning but I don't like it. I will look into how to use versioneer with setup.cfg for this.

dylanmcreynolds commented on setup.cfg:
```diff
@@ -4,3 +4,54 @@ style = pep440-post
+    pydantic
+    requests
+packages = find:
+python_requires  = >=3.8
```
This is good and correct, but the build is failing because the github action is [testing 3.7](https://github.com/SciCatProject/pyscicat/issues/36). Would you want to fix #36 in this PR?

dylanmcreynolds replied:
I really would like to match the version to those that CI tests.

dylanmcreynolds commented on pyscicat/tests/tests_local.py:
```diff
@@ -0,0 +1,78 @@
+from datetime import datetime
+import os
+
+class TestClientLocally(unittest.TestCase):
```
Do we really want to mix `uniitest` and `pytest` in the same project? I mean, it's handy, because CI uses pytest and this would break it, but it makes me a little  nervous to have multiple test frameworks like this.

LAShemilt replied:
you can use unittest and pytest together and it won't break CI, but I agree one framework should be used. These tests may disappear at some point, I wrote them initially to test the code with v4.

dylanmcreynolds reviewed (changes requested):
Thanks for this! I've never migrated a python project to use `setup.cfg`. I have some changes and questions.

LAShemilt commented:
Now finished work on the pyscicat version for interacting with the v4 scicat backend. Please see the following important changes to pyscicat:

**1) Install and Setup**
I have now split up the setup.py into a minimal setup.py script and a setup.cfg. This allows for easy configuration of unique setups which may require optional dependencies. To now install these please use for example:
`pip install -e ".[option]"` so to use the hd5f plugins install with `pip install -e ".hdf5"`

**2) Deprecation of methods**
I have used python's `deprecation` package to deprecate methods that no longer exist in the v4 backend. They will still work with the old backend, and can still be called with pyscicat as normal. However, you will get a deprecation warning when they are called.

**3) Endpoint routes**
I had really trouble with the getter methods which was coming from urllib's `urljoin`. If you use a `/` in an endpoint definition when you join that with a base url , `urljoin` will treat it as an absolute path and will append it after the port number. So `http://localhost:3000/api/v3`  joined with `/Datasets/` was becoming `http://localhost:3000/Datasets/`. This problem goes away when using proper routing as to not have the port in the url name, which is why I guess it has not affected production systems so far.  I have removed the `/` at the beginning so that the append will work properly in both the test and prod cases.

**4) Intergration testing in github actions**
I have updated the github actions to include some minor integration testing for the backend-v.4 . It uses the docker image of the last stable release from https://github.com/SciCatProject/scicat-backend-next. This runs as a service and the tests run a few methods against it to ingest and retrieve data. 

Any comments and review would be greatly appreciated.

dylanmcreynolds commented on pyscicat/client.py:
```diff
@@ -141,13 +143,18 @@ def _call_endpoint(
         )
         return result
 
+    @deprecated(deprecated_in='1.0.0', details='this method no longer works with the new Scicat '
```
Before we merge, let's talk about our desired release version.

nitrosx commented:
One more, given that this PR goes against the v4.x branch, should we go ahead and merge and than work on the v4.x to make it production ready?

dylanmcreynolds commented:
Tests are failing because of the change in how requirements are listed.

dylanmcreynolds commented:
Hmm, flake8 failures now.

LAShemilt commented:
Please see #40
